types/next-auth/providers.d.ts

```diff
@@ -1,56 +1,86 @@
-import { GenericObject } from './_utils';
+import { User } from '.';
+import { JWT } from './jwt';
+import { NonNullParams, NullableParams, WithAdditionalParams } from './_utils';
 
-export interface Providers {
+export interface Provider<T extends string | undefined = undefined, U = T extends string ? 'oauth' : string> {
+    id: T;
+    name: string;
+    type: U extends string ? U : 'oauth' | 'email' | 'credentials';
```

Comment:
Is there anyway we could simplify the generic here? I got a bit lost trying to follow the logic, I think a small comment explaining the branching would help 📐